File: FX Simulator/TradeViewController.m
//
//  TradeViewController.m
//  ForexGame
//
//  Created  on 2014/05/26.
//  
//

#import "TradeViewController.h"

#import <QuartzCore/QuartzCore.h>
#import "SaveData.h"
#import "ChartViewController.h"
#import "Market.h"
#import "ChartViewController.h"
#import "OrderManager.h"
#import "OrderResult.h"
#import "Rate.h"
#import "RatePanelViewController.h"
#import "SimulationManager.h"
#import "TradeDataViewController.h"

@interface TradeViewController ()

@end

@implementation TradeViewController {
    ChartViewController *_chartViewController;
    RatePanelViewController *_ratePanelViewController;
    TradeDataViewController *_tradeDataViewController;
    Market *_market;
    OrderManager *_orderManager;
    SaveData *_saveData;
    SimulationManager *_simulationManager;
    UIView *_adView;
}

- (void)loadSimulationManager:(SimulationManager *)simulationManager
{
    _simulationManager = simulationManager;
}

/**
 1回目は、prepareForSegueの前に呼ばれる
 */
- (void)loadSaveData:(SaveData *)saveData
{
    _saveData = saveData;
}

- (void)loadMarket:(Market *)market
{
    _market = market;
}

- (void)loadOrderManager:(OrderManager *)orderManager
{
    _orderManager = orderManager;
}

- (void)saveDataDidLoad
{
    if (!_chartViewController || !_ratePanelViewController || !_tradeDataViewController) {
        return;
    }
    
    [_chartViewController setChart:_saveData.mainChart];
    [_ratePanelViewController loadSaveData:_saveData];
    [_ratePanelViewController loadMarket:_market];
    [_ratePanelViewController loadOrderManager:_orderManager];
    [_tradeDataViewController loadSaveData:_saveData];
    [_tradeDataViewController loadMarket:_market];
    [_orderManager addDelegate:self];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if ([segue.identifier isEqualToString:@"ChartViewControllerSeg"]) {
        _chartViewController = segue.destinationViewController;
        _chartViewController.delegate = self;
    } else if ([segue.identifier isEqualToString:@"RatePanelViewControllerSeg"]) {
        _ratePanelViewController = segue.destinationViewController;
    } else if ([segue.identifier isEqualToString:@"TradeDataViewControllerSeg"]) {
        _tradeDataViewController = segue.destinationViewController;
        _tradeDataViewController.delegate = self;
    }
    
    [self saveDataDidLoad];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    if (!_simulationManager.isStartTime) {
        [_simulationManager startTime];
    } else {
        [_simulationManager resumeTime];
    }
    
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    
    [_simulationManager pauseTime];
}

/*-(void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event
{
    [tradeDataViewController tradeViewTouchesBegan];
}*/

- (void)autoUpdateSettingSwitchChanged:(BOOL)isSwitchOn
{
    [_simulationManager setIsAutoUpdate:isSwitchOn];
}

- (void)update
{
    [_chartViewController update:_market];
    [_ratePanelViewController update];
    [_tradeDataViewController update];
}

- (void)chartViewTouched
{
    [_simulationManager addTime];
}

- (void)didOrder:(OrderResult *)result
{
    [result completion:nil error:^{
        [result showAlertToController:self];
    }];
    
    [_tradeDataViewController didOrder:result];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: FX Simulator/TradeViewController.swift
//
//  TradeViewController.swift
//  FXSimulator
//
//  Created by yuu on 2016/04/27.
//
//

import UIKit

class TradeViewController: UIViewController {

    var chartViewController: ChartViewController!
    var ratePanelViewController: RatePanelViewController!
    var tradeDataViewController: TradeDataViewController!
    
    var market: Market!
    var orderFactory: OrderFactory!
    var orderManager: OrderManager!
    var saveData: SaveData!
    var simulationManager: SimulationManager!
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "ChartViewControllerSeg" {
            chartViewController = segue.destinationViewController as! ChartViewController
            chartViewController.delegate = self
        } else if segue.identifier == "RatePanelViewControllerSeg" {
            ratePanelViewController = segue.destinationViewController as! RatePanelViewController
        } else if segue.identifier == "TradeDataViewControllerSeg" {
            tradeDataViewController = segue.destinationViewController as! TradeDataViewController
            tradeDataViewController.delegate = self
        }
        
        self.saveDataDidLoad()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        if !simulationManager.isStartTime {
            self.update()
        } else {
            simulationManager.resumeTime()
        }
    }
    
    override func viewDidAppear(animated: Bool) {
        super.viewDidAppear(animated)
        
        if !simulationManager.isStartTime {
            simulationManager.startTime()
        }
    }
    
    override func viewWillDisappear(animated: Bool) {
        super.viewWillDisappear(animated)
        
        simulationManager.pauseTime()
    }
    
    func update() {
        chartViewController.update(market)
        ratePanelViewController.update()
        tradeDataViewController.update()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    

    /*
    // MARK: - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
    }
    */

}

extension TradeViewController: SimulationManagerDelegate {
    
    func loadSimulationManager(simulationManager: SimulationManager) {
        self.simulationManager = simulationManager
    }
    
    /**
     1回目は、prepareForSegueの前に呼ばれる
     */
    func loadSaveData(saveData: SaveData) {
        self.saveData = saveData
    }
    
    func loadMarket(market: Market) {
        self.market = market
    }
    
    func loadOrderFactory(orderFactory: OrderFactory) {
        self.orderFactory = orderFactory
    }
    
    func loadOrderManager(orderManager: OrderManager) {
        self.orderManager = orderManager
    }
    
    func saveDataDidLoad() {
        if chartViewController == nil || ratePanelViewController == nil || tradeDataViewController == nil {
            return
        }
        
        chartViewController.setChart(saveData.mainChart)
        
        ratePanelViewController.loadSaveData(saveData)
        ratePanelViewController.loadMarket(market)
        ratePanelViewController.loadOrderFactory(orderFactory)
        ratePanelViewController.loadOrderManager(orderManager)
        
        tradeDataViewController.loadSaveData(saveData)
        tradeDataViewController.loadMarket(market)
    }
    
    func marketDidUpdate() {
        self.update()
    }
    
    func simulationStopped(message: Message) {
        message.showAlertToController(self)
    }
    
    func didOrder(result: Result) {
        result.success(nil, failure: { result.showAlertToController(self) })
        
        tradeDataViewController.didOrder(result)
    }
}

extension TradeViewController: ChartViewControllerDelegate {
    
    func chartViewTouched() {
        simulationManager.addTime()
    }
}

extension TradeViewController: TradeDataViewControllerDelegate {
    
    var isAutoUpdate: Bool { return simulationManager.isAutoUpdate }
    
    func autoUpdateSettingSwitchChanged(isSwitchOn: Bool) {
        simulationManager.isAutoUpdate = isSwitchOn
    }
}
